Prelecture HW

In [5]:
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)
df.isna().sum()


# Display the columns and the size of the dataset
print("Columns in the dataset:", df.columns)
print("Number of rows and columns:", df.shape)


Columns in the dataset: Index(['row_n', 'id', 'name', 'gender', 'species', 'birthday', 'personality',
       'song', 'phrase', 'full_id', 'url'],
      dtype='object')
Number of rows and columns: (391, 11)


2.2. 
Observations are the individual data points/records in a dataset. In this dataset the observations would be the villagers from Animal Crossing. 
Variables are the different characteristics and information recorded for each observation, in this case for example, it is the name, gender, species, or personality of each Animal Crossing character. 

In [7]:
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)
df.isna().sum()

# Display the columns and the size of the dataset
print("Columns in the dataset:", df.columns)
print("Number of rows and columns:", df.shape)

# Provide summary for numeric columns
print("Summary for Numeric Columns:")
print(df.describe(), "\n")

# Provide summary for categorical columns
print("Summary for Categorical Columns:")
categorical_cols = df.select_dtypes(include=['object']).columns

for col in categorical_cols:
    print(f"\nColumn: {col}")
    print(df[col].value_counts())  # Display the frequency of unique values in each categorical column
    print("\n")

Columns in the dataset: Index(['row_n', 'id', 'name', 'gender', 'species', 'birthday', 'personality',
       'song', 'phrase', 'full_id', 'url'],
      dtype='object')
Number of rows and columns: (391, 11)
Summary for Numeric Columns:
            row_n
count  391.000000
mean   239.902813
std    140.702672
min      2.000000
25%    117.500000
50%    240.000000
75%    363.500000
max    483.000000 

Summary for Categorical Columns:

Column: id
id
admiral    1
mott       1
paula      1
patty      1
pate       1
          ..
eloise     1
elmer      1
ellie      1
elise      1
zucker     1
Name: count, Length: 390, dtype: int64



Column: name
name
Admiral    1
Muffy      1
Paula      1
Patty      1
Pate       1
          ..
Elvis      1
Eloise     1
Elmer      1
Ellie      1
Zucker     1
Name: count, Length: 391, dtype: int64



Column: gender
gender
male      204
female    187
Name: count, dtype: int64



Column: species
species
cat          23
rabbit       20
frog         18
squirrel     1

In [9]:
import pandas as pd

# Load the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

# Get the shape of the dataset
print("Shape of the dataset (rows, columns):", df.shape)

# Get the description of the dataset
print("\nDescription of numeric columns:")
print(df.describe())

Shape of the dataset (rows, columns): (891, 15)

Description of numeric columns:
         survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200


4)
a) In the titanic dataset, df.shape shows the total number of rows and columns in the dataset, and df.describe() shows the statistics for all the columns, however, by default it summarizes only numeric columns.
Based on df.shape showing (891,15), we can see that there are 891 rows and 15 columns. 
While .shape tells us that there are 15 columns, we can only see 6 columns from .describe(). 
There may be discrepancies between the number of columns reported in .shape and the values displayed from .describe() because .describe() only displays the summary statistics for numeric columns, while in the example of the titanic dataset below, non-numeric columns also exist. This is what led to the columns shown in .describe() to not match the number of columns in .shape. 
b) In .describe(), the "count" represents the number of non-missing entries in each column. If there are no missing values in the dataset, the count shown for the titanic dataset should be consistently 891. However, for the "count" for age column is 714 and not 891. Therefore, this reveals that in these columns 177 rows have missing values, which is why the "count" reported in .describe() is different from the rows reported in .shape. 

5) In programming, attributes have to do with the properties and characteristics of an object, and are written without parentheses, while methods are written to represent the actions that can be performed and are written with parentheses. Attributes give fixed information, but methods perform operations that return results. 

Link to chat log history 1: https://chatgpt.com/share/06af93c5-32e0-46d9-b66c-2069fc30bd3e
Summary of Chatbot session 1: 

"Sure! Here’s a summary of our conversation:

Dataset Overview:

You downloaded a dataset about Animal Crossing villagers and wanted to explore its structure.
We discussed the terms observations (rows) and variables (columns) in the context of your dataset.
Basic Dataset Exploration:

I provided code to load the dataset and display its columns, size (using .shape), and missing values (using .isna().sum()).
Generating Summaries:

I suggested using df.describe() for numeric data and df['column'].value_counts() for categorical data to generate simple summaries of the dataset.
I also explained how df.shape differs from df.describe() in terms of columns, as the latter only shows numeric columns by default.
Understanding Discrepancies:

We discussed the difference between row counts reported by .shape (total rows) and count in .describe() (non-missing values in specific columns).
Programming Concepts:

I explained the difference between an attribute (e.g., df.shape) and a method (e.g., df.describe()), with examples to clarify the concepts.
Overall, we explored dataset inspection, summarization techniques, and the difference between attributes and methods in programming. Let me know if you'd like further clarification or help with any topic!"

Postlecture HW

6)
Count: the number of non-missing values in the dataset for each column.
Mean: the average of values in a column.
Std: short for standard deviation, and measures the amount of deviation or difference between a set of values from the mean.
Min: the minimum or smallest value in the column.
25%: the first quartile of the distribution, the value 25% of the data falls below.
50%: the median/second quartile, showing the middle value of the data. This is the middle because half of the data is above it and half is below. 
75%: the third quartile, 75% of the data falls below it.
Max: the maximum or largest value in the column. It represents the maximum boundary of the data. 

In [8]:
import pandas as pd

# Sample DataFrame
data = {
    'price': [300000, 450000, None, 500000],
    'bedrooms': [3, 4, 3, None],
    'square_footage': [1500, 2000, None, 1800],
    'year_built': [2000, 2010, 2005, 2020],
    'renovated': [1, 0, 1, 1]
}
df = pd.DataFrame(data)

# Drop rows where 'price' or 'square_footage' is missing
df_cleaned = df.dropna(subset=['price', 'square_footage'])

7.1)
df.dropna() may be preferred when removing rows that have missing values in any columns. It removes a few rows of missing values but also keeps the rest of the data. 
del df['col'] is used when the entire column is unecessary and removing the entire column is preferred. In the example we drop the rows 'price' and 'square_footage' because it's missing important information that can influence the accuracy of the predictions. 

In [9]:
import pandas as pd

# Sample DataFrame
data = {
    'customer_id': [1, 2, 3, 4],
    'age': [25, 34, None, 29],
    'purchase_amount': [200, 150, 300, 250],
    'membership_status': ['Gold', 'Silver', 'Gold', 'Bronze'],
    'feedback_score': [None, 4, None, None]
}
df = pd.DataFrame(data)

# Drop the 'feedback_score' column
df_cleaned = df.drop(columns=['feedback_score'])

7.2)
del df['col'] is preferred because there are many missing values in the column 'feedback_score'. It would be better to just erase the column entirely instead of keeping the irrelevant information.

7.3)
When using df.dropna(), it helps drop any rows that have missing values by default, and this can lead to the loss of some important information by accident. By first using del df['col'] to drop a column with a lot of missing values, it helps df.dropna() focus on only the missing values in important columns without accidentally deleting useful information. 

7.4) From the titanic dataset, we can see that before we dropped anything we have missing data in the columns 'deck', 'age', 'embarked' and 'embark_town'. To remove all of the missing data I first went and considered the amount of missing data in the columns. Because the column 'deck' has a large amount of missing data, I decided to use del df['col'] and drop the column entirely before moving on to using df.dropna() to avoid deleting any more important information. I use df.dropna() to remove the small amounts of missing data from 'age', 'embarked' and 'embark_town'to finally reach the after results. 

In [56]:
#Before

import pandas as pd

# Load the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

# Displays the entire DataFrame
print(df)

# Check for missing data in the Titanic dataset
missing_data = df.isnull().sum()

# Display the missing data
print(missing_data)

     survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alo

In [57]:
#After

import pandas as pd

# Load the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

# Displays the entire DataFrame
print(df)

del df['deck']

df = df.dropna(subset=['age', 'embarked'])


# Check for missing data in the Titanic dataset
missing_data = df.isnull().sum()

# Display the missing data
print(missing_data)

     survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alo

Link to chat log history 2:https://chatgpt.com/share/f1d76e34-942c-4457-a3d3-6277ed0d0412
Summary of Chatbot session 2: "Here’s a summary of our chatbot session so far:

Summary Statistics Definitions: I provided definitions for the summary statistics shown in the df.describe() method, including count, mean, std, min, 25%, 50%, 75%, and max.

Use Case of df.dropna() vs del df['col']: I gave an example where df.dropna() is preferred over del df['col'] when cleaning a dataset by removing rows with missing values in critical columns.

Opposite Use Case: I provided a reverse example where using del df['col'] is preferred over df.dropna(), particularly when removing an entire irrelevant column with many missing values to retain as many rows as possible.

Order of Applying del df['col'] and df.dropna(): I explained why it's important to use del df['col'] before df.dropna()—to avoid unnecessary data loss when dealing with irrelevant columns that contain missing values.

Accessing the Titanic Dataset: You asked for help accessing the entire Titanic dataset. I showed you how to load the dataset and view it entirely, as well as how to use options to display all rows and columns.

Checking for Missing Data: You wanted to check for missing data in the Titanic dataset. I tried to run a Python tool to check this, but I wasn’t able to execute it. I provided code you can use to check missing data on your own.

Let me know if you need anything else!"

Link to chat log history 3: https://chatgpt.com/share/a33aa2b0-af75-4b09-9cbc-46c2bc4a3364
Summary of Chatbot session 3:
"Sure! Here’s a summary of our chat:

Objective: You wanted to manipulate the Titanic dataset by either dropping columns or rows.

Initial Code: You provided code to load the Titanic dataset using pandas and requested help with modifying it.

Dropping Columns: I showed you how to drop columns 'age' and 'embarked' from the DataFrame using df.drop(columns=['age', 'embarked']).

Dropping Rows: After clarifying that you wanted to remove rows instead of columns, I provided code to drop rows where either 'age' or 'embarked' is missing using df.dropna(subset=['age', 'embarked']).

Let me know if there’s anything else you need!"

8.1)
By typing df.groupby("col1") the dataframe is grouped by the values in column one, then adding ["col2"].describe() to it we are selecting column two to describe and summarize statistics for column two grouped by the values in column one. 

In [11]:
import pandas as pd

# Load the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

print(df.head())
print(df.columns)

# Group by 'class' and describe 'age'
grouped_description = df.groupby("class")["age"].describe()

print(grouped_description)

   survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True  
Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'class', 'who', 'adult_male', 'deck', 'embark_town',
       'alive', 'alone'],
      dtype='object')
        count       mean      

8.2)
df.describe() shows the count for the entire dataframe over all rows, and ignores missing values, giving us overall number of non-missing values in each column. However, in df.groupby("col1")["col2"].describe() the count is represented separately for each group, meaning the count will show the missing values in column two within each group in column one. Since df.describe() gives an overall count, when there is a lower count it reflects that the column has missing values, but in a grouped count such as df.groupby("col1")["col2"].describe() some groups may have no missing values, resulting in a count equal to the number of rows in the groups. 

Link to chat log history 4: https://chatgpt.com/share/4c94ccf6-f6ea-4553-90f0-d92dbbb5d1f4
Summary of Chatbot session 4: "Sure! Here’s a summary of our session:

Understanding df.groupby("col1")["col2"].describe():

This method is used to get descriptive statistics for a column (col2) grouped by another column (col1).
We demonstrated this using the Titanic dataset, grouping by class and describing the age column.
Difference in count Values:

df.describe() provides overall statistics for each column, with count reflecting the number of non-missing values across the entire DataFrame.
df.groupby("col1")["col2"].describe() provides descriptive statistics for col2 within each group defined by col1, with count reflecting non-missing values within each specific group.
The counts might differ because df.describe() aggregates over the whole DataFrame, while the grouped describe() calculates statistics separately for each group, which might have different counts based on missing values.
Let me know if there’s anything more you’d like to explore or clarify!"

8.3) A)

In [1]:
# Load the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

print(df.head())
print(df.columns)

# Group by 'class' and describe 'age'
grouped_description = df.groupby("class")["age"].describe()

print(grouped_description)

NameError: name 'pd' is not defined

8.3) B)

In [5]:
import pandas as pd

# Load the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanics.csv"
df = pd.read_csv(url)

print(df.head())
print(df.columns)

# Group by 'class' and describe 'age'
grouped_description = df.groupby("class")["age"].describe()

print(grouped_description)

HTTPError: HTTP Error 404: Not Found

8.3) C)

In [8]:
import pandas as pd

# Load the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

print(df.head())
print(df.columns)

# Group by 'class' and describe 'age'
grouped_description = DF.groupby("class")["age"].describe()

print(grouped_description)

   survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True  
Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'class', 'who', 'adult_male', 'deck', 'embark_town',
       'alive', 'alone'],
      dtype='object')


NameError: name 'DF' is not defined

8.3) D)

In [10]:
import pandas as pd

# Load the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

print(df.head()
print(df.columns)

# Group by 'class' and describe 'age'
grouped_description = df.groupby("class")["age"].describe()

print(grouped_description)

SyntaxError: '(' was never closed (3911219227.py, line 7)

8.3) E)

In [11]:
import pandas as pd

# Load the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

print(df.head())
print(df.columns)

# Group by 'class' and describe 'age'
grouped_description = df.groopby("class")["age"].describe()

print(grouped_description)

   survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True  
Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'class', 'who', 'adult_male', 'deck', 'embark_town',
       'alive', 'alone'],
      dtype='object')


AttributeError: 'DataFrame' object has no attribute 'groopby'

8.3) F)

In [12]:
import pandas as pd

# Load the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

print(df.head())
print(df.columns)

# Group by 'class' and describe 'age'
grouped_description = df.groupby("Class")["age"].describe()

print(grouped_description)

   survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True  
Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'class', 'who', 'adult_male', 'deck', 'embark_town',
       'alive', 'alone'],
      dtype='object')


KeyError: 'Class'

8.3) G)

In [14]:
import pandas as pd

# Load the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

print(df.head())
print(df.columns)

# Group by 'class' and describe 'age'
grouped_description = df.groupby("class")[age].describe()

print(grouped_description)

   survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True  
Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'class', 'who', 'adult_male', 'deck', 'embark_town',
       'alive', 'alone'],
      dtype='object')


NameError: name 'age' is not defined

8.3) I think overall its definitely easier to work with a Chatbot to help with the code errors. ChatGPT replied very quickly and could identify the issue with the code, while going through Google was more of a hassle looking for the solution. There is also no guarantee you can find the solution to your problem even when spending more time on Google.

Link to chat log history 5: https://chatgpt.com/share/7b5e37e5-856b-4cc1-9444-c6433cdef548
Summary of Chatbot session 5: "Certainly! Here's a summary of the chatbot session:

Initial Issue:

You encountered a NameError due to not importing pandas. This was resolved by importing pandas as pd.
Dataset Loading Error:

You received an HTTPError because of a typo in the dataset URL ("titanics.csv" instead of "titanic.csv"). Correcting the URL fixed the issue.
Variable Name Case-Sensitivity:

You encountered a NameError due to using DF instead of df. This was resolved by ensuring consistent use of the correct variable name.
Syntax Error:

There was a syntax error because a closing parenthesis was missing in the print(df.head() function. Adding the missing parenthesis fixed it.
Attribute Error:

You encountered an AttributeError due to a typo in groopby instead of groupby. This was corrected.
KeyError:

You faced multiple KeyError issues due to incorrect capitalization of column names, specifically using "Class" instead of "class". This was corrected by using the correct column names.
NameError:

Finally, you faced a NameError when trying to reference age without quotes. This was resolved by placing "age" in quotes as it represents the column name.
Throughout the session, we identified and resolved issues related to typos, case sensitivity, and proper column name usage, allowing the code to execute successfully."

9) Somewhat, I will try my best ;-;